odl/tomo/operators/xray_trafo.py

```diff
@@ -0,0 +1,241 @@
+            Implementation backend for the transform. Supported backends:
+            'astra': ASTRA toolbox, uses CPU or CUDA depending on the
+            underlying data space of ``discr_dom``
+            'astra_cpu': ASTRA toolbox using CPU, only 2D
+            'astra_cuda': ASTRA toolbox, using CUDA, 2D or 3D
+            Default: 'astra'
+        kwargs : {'range_interpolation'}
+            'range_interpolation' : {'nearest', 'linear', 'cubic'}
+                Interpolation type for the discretization of the
+                operator range.
+                Default: 'nearest'
+        """
+        if not isinstance(discr_dom, DiscreteLp):
+            raise TypeError('discretized domain {!r} is not a `DiscreteLp`'
+                            ' instance.'.format(discr_dom))
+
+        if not isinstance(geometry, Geometry):
+            raise TypeError('geometry {!r} is not a `Geometry` instance.'
+                            ''.format(geometry))
+
+        if not (geometry.has_motion_sampling and geometry.has_det_sampling):
+            raise ValueError('geometry {} does not have sampling grids for '
+                             'both motion and detector.'.format(geometry))
+
+        backend = str(backend).lower()
+        if backend not in _SUPPORTED_BACKENDS:
+            raise ValueError('backend {!r} not supported.'
+                             ''.format(backend))
+
+        if backend == 'astra':
+            if not ASTRA_AVAILABLE:
+                raise ValueError('ASTRA backend not available.')
+            if (isinstance(discr_dom.dspace, CudaNtuples) and
+                    not ASTRA_CUDA_AVAILABLE):
+                raise ValueError('ASTRA CUDA backend not available.')
+            if discr_dom.dspace.dtype not in (np.float32, np.complex64):
+                raise ValueError('ASTRA support is limited to `float32` '
+                                 'for real and `complex64` for complex '
+                                 'data.')
+            if not np.allclose(discr_dom.grid.stride[1:],
+                               discr_dom.grid.stride[:-1]):
+                raise ValueError('ASTRA does not support different pixel/voxel'
+                                 ' sizes per axis (got {}).'
+                                 ''.format(discr_dom.grid.stride))
+
+        self._geometry = geometry
+
+        # Create discretized space (operator range). Use same data space
+        # type as the domain.
+        # TODO: maybe use a ProductSpace structure
+        ran_uspace = FunctionSpace(geometry.params)
+        # CHECKME: Is this the right weight?
+        ran_dspace = discr_dom.dspace_type(geometry.grid.ntotal,
+                                           weight=geometry.grid.cell_volume,
+                                           dtype=discr_dom.dspace.dtype)
+
+        ran_interp = kwargs.pop('range_interpolation', 'nearest')
+        discr_ran = DiscreteLp(ran_uspace, geometry.grid, ran_dspace,
```

Comment:
For practicability, I revise my first comment partly and suggest that we (generally) follow two principles, namely that we

- don't do mathematical bullshit ourselves,
- don't stand in the way of users who want to do bullshit.